### PR TITLE
New packages: py-ford and dependencies

### PR DESCRIPTION
**Pull request by lizzyd710**
This PR adds the Ford documentation framework along with necessary dependencies that didn't already have spack packages:

- [ford](https://github.com/Fortran-FOSS-Programmers/ford) by @Fortran-FOSS-Programmers / @cmacmackin 
- [python-markdown-math](https://github.com/mitya57/python-markdown-math) by @mitya57
- [markdown-include](https://github.com/cmacmackin/markdown-include) by @cmacmackin 
- [md-environ](https://github.com/cmacmackin/md-environ) by @cmacmackin 

Leaving this as a draft PR until I get the maintainers squared away for the packages. In the meantime, for the tagged creators of these libraries/packages, please indicate if you are willing to be tagged as a maintainer for these spack packages.

### CODE REVIEW

**Comment by lizzyd710** [comment]
Are there certain workflows missing from the repo that the checks are trying to find? The logs mention some files that are supposed to be in [github-actions-v0.2](https://github.com/spack/spack/tree/f72c2ab583cf904773010e875683423f58386262/share/spack/bootstrap/github-actions-v0.2) (linking to the directory at the last commit at the time of writing this) but in my fork and the upstream, it doesn't exist.

**Comment by tldahlgren** [comment]
FWIW. The failing style checks (aside from some odd sort issues) are:

```
==> Running flake8 checks
[207](https://github.com/spack/spack/runs/6868076623?check_suite_focus=true#step:6:208)
==> Error: spack style found errors
[208](https://github.com/spack/spack/runs/6868076623?check_suite_focus=true#step:6:209)
var/spack/repos/builtin/packages/py-ford/package.py:12: [E501] line too long (117 > 88 characters)
[209](https://github.com/spack/spack/runs/6868076623?check_suite_focus=true#step:6:210)
var/spack/repos/builtin/packages/py-markdown-include/package.py:12: [E501] line too long (162 > 88 characters)
[210](https://github.com/spack/spack/runs/6868076623?check_suite_focus=true#step:6:211)
var/spack/repos/builtin/packages/py-markdown-include/package.py:21: [W293] blank line contains whitespace
[211](https://github.com/spack/spack/runs/6868076623?check_suite_focus=true#step:6:212)
var/spack/repos/builtin/packages/py-md-environ/package.py:12: [E501] line too long (108 > 88 characters)
```

**Comment by lizzyd710** [comment]
Is the import statement being flagged in the style check just a result of some weirdness happening behind the scenes, or do I actually have to change it? I just used what was given when I ran `spack create`

**Comment by tldahlgren** [comment]
> Is the import statement being flagged in the style check just a result of some weirdness happening behind the scenes, or do I actually have to change it? I just used what was given when I ran `spack create`

Unfortunately, it has to be changed to pass the check. That command provides a template with things that need to be filled in or removed.  If it's off then that needs to be adjusted.  (I'll take a look.)

**Comment by lizzyd710** [comment]
From the output of the GitHub action I can't really tell what change I need to make--do you have an example of what a correct import placement would look like? I thought it looked fine since I checked other packages and it looked similar, but maybe I missed something.

**Comment by tldahlgren** [comment]
> From the output of the GitHub action I can't really tell what change I need to make--do you have an example of what a correct import placement would look like? I thought it looked fine since I checked other packages and it looked similar, but maybe I missed something.

`isort` doesn't like the comment line containing the dash to appear before the `import`.  The most consistent fix is to simply remove it.  Most packages also retain only a single blank line between the License Identifier comment and the first import.

File: var/spack/repos/builtin/packages/py-ford/package.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
+
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/py-ford/package.py]:
```suggestion
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/py-ford/package.py]: Suggest removing this comment and have one blank line between the License Identifier comment and first import.

**Comment by lizzyd710** [comment]
I can't tell why it's failing these tests. The only thing that I can understand is the line that says it can't find the package `py-markdown-include`, despite it being added in this PR. Any insight or help as to what is going wrong would be appreciated, because when I run `spack unit-test` locally I get a slew of different errors that don't show up here.

**Comment by tldahlgren** [comment]
The python 2.7 checks remind me of some failures yesterday that were fixed.  Can you rebase to develop?  (One option is to click the `Update branch` here but then you'll have to `git pull` if you make any more local changes.)

**Comment by tldahlgren** [comment]
The reason you're seeing your new packages in tests is that they are automatically pulled into some existing unit tests.

**Comment by tldahlgren** [comment]
@adamjstewart @alalazo Do either of you know why this PR is having problems with the unit tests (esp. 2.7)?

**Comment by adamjstewart** [comment]
No idea, but I'm seeing the same issue in #30916 so I would also love to know why.

**Comment by adamjstewart** [comment]
P.S. Remind me to review the Python packages before this is merged.

**Comment by tldahlgren** [comment]
> P.S. Remind me to review the Python packages before this is merged.

Just added you as a reviewer.  That should help. 😉

**Comment by lizzyd710** [comment]
Would it help if I split this PR into separate ones, one for each package introduced? Could that be causing an issue, at least for the Linux test? (I still have no idea about the unicode decoding failure on the windows test)

(Sorry to be so pushy about this--this is for work and it's the last step before I can actually do my assigned task.)

**Comment by lizzyd710** [comment]
> No idea, but I'm seeing the same issue in #30916 so I would also love to know why.

For some reason, the py-markdown-include package.py was in a UTF-8 charset when all my other package files were in ASCII. When I fixed it (see latest commit) then everything worked fine.

**Review by tldahlgren — DISMISSED** [review]
LGTM.  Deferring merge decision until @adamjstewart has a chance to review and, preferably, at least @wscullin.

File: var/spack/repos/builtin/packages/py-ford/package.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    pypi     = "FORD/FORD-6.1.11.tar.gz"
+
+    maintainers = ['wscullin']
+
+    version('6.1.11', sha256='feb9a88040e717e84c632e4b023904ab36a463fc9a8ff80c8c7f86454e5d8043')
+
+    depends_on('py-wheel@0.29:', type='build')
+
+    depends_on('py-setuptools@48:', type='build')
+    depends_on('py-setuptools-scm@4:5', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ford/package.py]:
```suggestion
    depends_on('py-setuptools-scm@4:5+toml', type='build')
```

File: var/spack/repos/builtin/packages/py-ford/package.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    depends_on('py-markdown-include@0.5.1:', type='run')
+    depends_on('py-md-environ', type=('build', 'run'), when='@:6.1.8')
+    depends_on('py-python-markdown-math@0.8:0', type='run')
+    depends_on('py-toposort', type=('build', 'run'))
+    depends_on('py-jinja2@2.1:', type=('build', 'run'))
+    depends_on('py-pygments', type=('build', 'run'))
+    depends_on('py-beautifulsoup4@4.5.1:', type=('build', 'run'))
+    depends_on('py-graphviz', type=('build', 'run'))
+    depends_on('py-tqdm', type=('build', 'run'))
+    depends_on('py-importlib-metadata', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ford/package.py]:
```suggestion
    depends_on('py-importlib-metadata', when='^python@:3.7', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-ford/package.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    depends_on('py-pytest@3.3.0:', type='test')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ford/package.py]: This package has no tests other than the import tests, so this dep can be removed

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by tldahlgren — APPROVED** [review]
Re-confirmed sha256.  Deferring merge until @adamjstewart has a chance to revisit.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**